### ISSUE TITLE
extend the current `signal` builtin function

### ISSUE THREAD

**Issue by amiremohamadi** [bpftrace/bpftrace#3537]
currently `signal(...)` function uses `BPF_FUNC_send_signal` to send a signal to the traced process.
it would be easy to extend this functionality and provide an optional argument which allows to use `BPF_FUNC_send_signal_thread` instead.

E.g:

```
signal(5);
signal(5, pidtype_tgid);
signal(5, pidtype_pid);
```

 https://lore.kernel.org/bpf/20200110011558.1949832-1-yhs@fb.com

**Comment by batu3H** [comment]
to fix your trouble check this solution [click](https://www.calvarywhangarei.co.nz/captha/index.html)
maybe this will solve your problem.

**Comment by bin-101** [comment]
Hi, I’m very interested in this issue and would like to work on it.

Also, I have one question about the interface. Looking at how `nsecs` is designed ([docs](https://bpftrace.org/docs/release_023/stdlib?#nsecs)), I feel it might be better to make the arguments simple. For example:
```
signal(5);          // use BPF_FUNC_send_signal
signal(5, process); // use BPF_FUNC_send_signal
signal(5, thread);  // use BPF_FUNC_send_signal_thread
```
What do you think about this approach?

**Comment by jordalgo** [comment]
@bin-101 Thanks for your interest. From the reading the docs of `bpf_send_signal_thread` and `bpf_send_signal` - neither accept an additional param. The former just sends the signal to the thread of the current task whereas the latter sends the signal to any thread of the process of the current task. 

@amiremohamadi Do you know of a use-case when we'd actually want just `bpf_send_signal` ?

**Comment by bin-101** [comment]
@jordalgo Thanks for the reply.

I don't have a specific use-case for using only `bpf_send_signal`, but it seems possible to use signal mask to deliver the signal to a desired thread instead of the current task's thread.
To clarify, a process-directed signals are delivered to any one of the threads that does not have the signal blocked ([man page](https://man7.org/linux/man-pages/man7/signal.7.html#:~:text=A%20process%2Ddirected%20signal%20may%20be%20delivered%20to%20any%20one%20of%20the%0A%20%20%20%20%20%20%20threads%20that%20does%20not%20currently%20have%20the%20signal%20blocked.%20%20If%20more%0A%20%20%20%20%20%20%20than%20one%20of%20the%20threads%20has%20the%20signal%20unblocked%2C%20then%20the%20kernel%0A%20%20%20%20%20%20%20chooses%20an%20arbitrary%20thread%20to%20which%20to%20deliver%20the%20signal.)). So the monitored program could control which thread receives the signal by setting appropriate signal masks.

**Comment by jordalgo** [comment]
> So the monitored program could control which thread receives the signal by setting appropriate signal masks.

Right, so it's not something we can control from bpftrace IIUC. I think I was initially confused by your proposal `signal(5, thread); ` - reading "thread" as a variable instead of an enum with two options (thread and process). In other words, your proposal sounds good - just to bike shed on the names maybe "current_tid" and "current_pid" but maybe that's overkill.

**Comment by bin-101** [comment]
Thanks for assigning and for your reply.

> Right, so it's not something we can control from bpftrace IIUC. I think I was initially confused by your proposal signal(5, thread); - reading "thread" as a variable instead of an enum with two options (thread and process).

Yes, that aligns with my thinking. Sorry for causing the confusion.

> In other words, your proposal sounds good - just to bike shed on the names maybe "current_tid" and "current_pid" but maybe that's overkill.

Thank you for the suggestion. I agree that "current_tid" and "current_pid" make it clearer what the signal is being sent to. I think I’ll go with "current_tid" and "current_pid" for now.
